src/pages/payments/modules/payments/PaymentsContext.tsx

```diff
@@ -0,0 +1,33 @@
+import { ActionType } from "pages/payments/modules/payments/PaymentsActionType";
+import PaymentsReducer, {
+  defaultValue,
+  DefaultValueState,
+} from "pages/payments/modules/payments/PaymentsReducer";
+import { createContext, Dispatch, useContext, useReducer } from "react";
+
+export const PaymentsStateContext =
+  createContext<DefaultValueState>(defaultValue);
+
+type DispatchContext = Dispatch<ActionType>;
+export const PaymentsDispatchContext = createContext<DispatchContext>(
+  () => null
+);
+
+export const PaymentsProvider = ({
+  children,
+}: {
+  children: React.ReactNode;
+}) => {
+  const [state, dispatch] = useReducer(PaymentsReducer, defaultValue);
+
+  return (
+    <PaymentsStateContext.Provider value={state}>
+      <PaymentsDispatchContext.Provider value={dispatch}>
+        {children}
+      </PaymentsDispatchContext.Provider>
+    </PaymentsStateContext.Provider>
+  );
+};
+
+export const usePaymentsState = () => useContext(PaymentsStateContext);
+export const usePaymentsDispatch = () => useContext(PaymentsDispatchContext);
```

Comment:
이러한 훅은 실수로 Context 외부에서 사용할 여지가 있는데요.

따로 Error 처리를 해주시면 보다 안전하게 사용이 가능할거 같네요.
```suggestion
export const usePaymentsState = () => {
  const context = useContext(PaymentsStateContext);

  if (!context) {
    console.Error('usePaymentsState는 PaymentsProvider 안에서만 사용 가능합니다.');
  }

  return context;
};
```

Reply:
위의 코드는 예시일 뿐이니 어떻게 에러 핸들링을 하실 지는 유정님이 고민해보셨으면 좋겠네요! 👍

Reply:
``[bddb288](https://github.com/next-step/react-payments/pull/79/commits/bddb2881973395c18e3ace259de721a58287d7b0) 리액트의 에러바운더리 컴포넌트를 활용하는 방법으로 에러 핸들링을 구현해보았습니다! 실제 Context가 외부에서 사용하려 했을 때 에러가 나는지 확인을 해보고자 아래 `ErrorTest.tsx`파일 내에서 `usePaymentsState`에 접근하는 코드를 적어보았었습니다.

```typescript

function App() {
  return (
    <div className="App">
      <ErrorBoundary>
        <PaymentsProvider>
          <RouterProvider router={router} />
        </PaymentsProvider>

        **<ErrorText />**
      </ErrorBoundary>
    </div>
  );
}

import { usePaymentsState } from "./payments/modules/payments/PaymentsContext";

const ErrorTest = () => {
  const data = usePaymentsState();
  console.log(data);
  return (
    <div>
      ErrorTest 컴포넌트 입니다. <span>{data.selectedCard?.name}</span>
    </div>
  );
};

export default ErrorTest;
```

Reply:
``[bddb288](https://github.com/next-step/react-payments/pull/79/commits/bddb2881973395c18e3ace259de721a58287d7b0) 리액트의 에러바운더리 컴포넌트를 활용하는 방법으로 에러 핸들링을 구현해보았습니다! 실제 Context가 외부에서 사용하려 했을 때 에러가 나는지 확인을 해보고자 아래 `ErrorTest.tsx`파일 내에서 `usePaymentsState`에 접근하는 코드를 적어보았었습니다.

```typescript

function App() {
  return (
    <div className="App">
      <ErrorBoundary>
        <PaymentsProvider>
          <RouterProvider router={router} />
        </PaymentsProvider>

        **<ErrorText />**
      </ErrorBoundary>
    </div>
  );
}

/** ErrorTest **/
import { usePaymentsState } from "./payments/modules/payments/PaymentsContext";

const ErrorTest = () => {
  const data = usePaymentsState();

  return (
    <div>
      ErrorTest 컴포넌트 입니다. <span>{data.selectedCard?.name}</span>
    </div>
  );
};

export default ErrorTest;
```

위와 같이 처음 코드를 짜게 됐었는데 바로 에러를 던질거라 생각했던 제 처음 생각과는 달리 에러를 내뱉지 않았습니다. 이에 의문점이 생겨서 `console.log(data)`를 찍어보니 아래와 같이 deafultValue값이 나오게 되는 것을 확인할 수 있었습니다.
<img width="917" alt="image" src="https://user-images.githubusercontent.com/61952198/223730833-2187ecf0-f424-40d2-9015-3e4095b2fd0f.png">

이 부분에 대해서 제가 놓치고 있던 부분이 있었던 것 같아 [공식문서를 다시 살펴보니 ](https://reactjs.org/docs/context.html#reactcreatecontext) 매칭되는 Provider가 존재하지 않을 때에는 defaultValue가 사용된다라는 부분이 있었습니다. 

defaultValue를 null로 설정하고서 같은 코드를 실행하니 제대로 error를 throw하는 것을 확인 할 수 있었습니다. 
`export const PaymentsStateContext =
  createContext<DefaultValueState | null>(null);`
<img width="1049" alt="image" src="https://user-images.githubusercontent.com/61952198/223732625-4a562d83-6739-48a8-959c-fae8ed6334ab.png">


그러면서 한가지 궁금증이 생긴 부분이 있었는데요 ( 상황을 말씀 드리다보니 앞에 이야기가 조금 길어졌네요! )
에러핸들링을 하는 부분을 추가 하게 된다면 에러핸들링을 하고자 하는 부분에서 의도한대로 에러 처리가 이뤄져야 하는것이 아닐까 하는 생각이 들었습니다. defaultValue를 통한 방법도 Context상에서의 또 하나의 에러를 다루는 방법일 수 있겠지만 실수로 hook을 Context 밖에서 사용하는것을 예방하기 위해서는 좀 더 직관적으로 보여질 수 있는 에러바운더리로 처리해야하는가 하는 여러 고민점들이 있었습니다. 이러한 상황에 대해서 윤호님의 생각은 어떠신지 궁금합니다! (어느것이 좋다하기 보다는 어떤 상황인지에 따라 처리하는 방식이 달라지게 될까요?)

Reply:
상황에 따라 달라지겠지만 저 역시 ErrorBoundary를 사용하는걸 선호합니다!

단순히 throw하게 되면 catch하는 부분이 없을 때 애플리케이션 전체가 사용불가능한 상황이 생길 수 있는데요.

ErrorBoundary를 사용하게 되면 부분적으로만 사용이 불가능하기 때문에 사용성 측면에서도 ErrorBoundary가 더욱 뛰어나다고 생각합니다.

그러나 지금은 Application 전체가 하나의 Context를 공유해서 이러한 효과까지는 누리기 어려울 수 있지만요